lib/elixir/lib/application.ex

```diff
@@ -412,12 +439,154 @@ defmodule Application do
+  def put_compile_env_listener(pid) when is_pid(pid) do
+    :elixir_config.update(:compile_env_listeners, &[pid | &1])
+    :ok
+  end
+
+  @doc """
+  Deletes a `pid` from compile env listeners.
+
+  See `get_compile_env_listeners/0` for more information.
+  """
+  @doc since: "1.10.0"
+  @spec delete_compile_env_listener(pid) :: :ok
+  def delete_compile_env_listener(pid) when is_pid(pid) do
+    :elixir_config.update(:compile_env_listeners, &List.delete(&1, pid))
+    :ok
+  end
+
+  @doc """
+  Reads the application environment at compilation time.
+
+  Similar to `get_env/3`, except it must be used to read values
+  at compile time. This allows Elixir to track when configuration
+  values between compile time and runtime differ.
+
+  The first argument is the application name. The second argument
+  `key_or_path`, is either an atom key or a path to traverse in
+  search of the configuration, starting with an atom key.
+
+  For example, imagine the following configuration:
+
+      config :my_app, :key, [foo: [bar: :baz]]
+
+  We can access it during compile time as:
+
+      Application.compile_env(:my_app, :key)
+      #=> [foo: [bar: :baz]]
+
+      Application.compile_env(:my_app, [:key, :foo])
+      #=> [bar: :baz]
+
+      Application.compile_env(:my_app, [:key, :foo, :bar])
+      #=> :baz
+
+  A default value can also be given as third argument. If
+  any of the keys in the path along the way is missing, the
+  default value is used:
+
+      Application.compile_env(:my_app, [:unknown, :foo, :bar], :default)
+      #=> :default
+
+      Application.compile_env(:my_app, [:key, :unknown, :bar], :default)
+      #=> :default
+
+      Application.compile_env(:my_app, [:key, :foo, :unknown], :default)
+      #=> :default
+
+  Giving a path is useful to let Elixir know that only certain paths
+  in a large configuration are compile time dependent.
+  """
+  # TODO: Warn if get_env/fetch_env/fetch_env! is used at compile time instead of compile_env
+  @doc since: "1.10.0"
+  @spec compile_env(app, key, value) :: value
+  def compile_env(app, key_or_path, default \\ nil) when is_atom(app) do
+    case fetch_compile_env(app, key_or_path) do
+      {:ok, value} -> value
+      :error -> default
+    end
+  end
+
+  @doc """
+  Reads the application environment at compilation time or raises.
+
+  This is the same as `compile_env/3` but it raises an
+  ArgumentError if the configuration is not available.
+  """
+  @doc since: "1.10.0"
+  @spec compile_env!(app, key) :: value
+  def compile_env!(app, key_or_path) when is_atom(app) do
+    case fetch_compile_env(app, key_or_path) do
+      {:ok, value} ->
+        value
+
+      :error ->
+        raise ArgumentError,
+              "could not fetch application environment #{inspect(key_or_path)} for application " <>
+                "#{inspect(app)} #{fetch_env_failed_reason(app, key_or_path)}"
+    end
+  end
+
+  defp fetch_compile_env(app, key) when is_atom(key),
```

Comment:
Shouldn't this be public to mirror fetch vs get for the normal env?

Reply:
No plans for now to make it public. We will see. :)